### PR TITLE
make TakeCell::empty() const, and make nrf52840 GPIO pin initialization a const fn

### PR DESCRIPTION
**Pull request by folkertdev**
### Pull Request Overview

This pull request makes the `TakeCell::new` function const. This requires an associated constant: moving the constant into `fn new` does not work, not using an explicit constant also does not work.

That function being const can then be used to initialize the gpio pins for the ntrf52840 in a `const fn`. It can't happen in an actual const because the final type contains a (non-static) lifetime. 

There are many more functions that can now become a `const fn`: I can add some to this PR but those are scattered all around the codebase.

### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make prepush`.

### CODE REVIEW

**Comment by bradjc** [comment]
This effectively reverts 4a465e6360e11f28d2bf75b5c6bd2d60cde77811, and implements #2214.

I assume this is due to improvements in the (stable) Rust compiler. I think this is a change we want, but wanted to link the history.

**Comment by folkertdev** [comment]
This trick has actually worked for a long time (apparently, since 1.32 https://godbolt.org/z/5bWhYf5fP), but only because the const is defined as an associated constant. The mutable reference in a const error is triggered when you put the const elsewhere (e.g. in the body of `new`)

**Review by brghena — DISMISSED** [review]

**Comment by brghena** [comment]
Bringing to the attention of @lschuermann

File: chips/nrf52840/src/interrupt_service.rs [hunk 1]
```diff
@@ -18,7 +18,7 @@ pub struct Nrf52840DefaultPeripherals<'a> {
 }
 
 impl<'a> Nrf52840DefaultPeripherals<'a> {
-    pub unsafe fn new(
```
Comment (hudson-ayers) [inline chips/nrf52840/src/interrupt_service.rs:21]: I suspect this change is something we want, but I don't immediately recall the rationale for why these were made unsafe (something related to multiple instantiations of the peripheral registers..?) so I would prefer this not be thrown into this PR. Instead we should have a seperate PR to remove this from the ~8 chips that have unsafe constructors for their `DefaultPeripherals`

**Review by hudson-ayers — COMMENTED** [review]
One comment, otherwise looks good

**Review by hudson-ayers — APPROVED** [review]

**Review by lschuermann — APPROVED** [review]
Sorry for my late response; this looks good to me. I was pretty sure that this would still allow creation of trait objects due to this associated const not being part of the trait, but wanted to double check. This still works, so we won't run into any issues in the future here!

**Pull request merged**